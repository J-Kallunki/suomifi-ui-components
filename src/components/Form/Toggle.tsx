import React, {
  Component,
  ReactElement,
  ReactNode,
  ComponentClass,
  FunctionComponent,
} from 'react';
import { HtmlLabel, HtmlInput } from '../../reset';
import { HtmlSpan } from '../../reset/HtmlSpan/HtmlSpan';
import classnames from 'classnames';

const baseClassName = 'fi-toggle';
const toggleDisabledClassName = `${baseClassName}--disabled`;
const toggleInputClassName = `${baseClassName}-input`;
const toggleLabelClassName = `${baseClassName}-label`;

export interface ToggleInputProps {
  /** State of input checkbox */
  checked?: boolean;
  /** Custom classname to extend or customize */
  className?: string;
  /** Disable Button usage */
  disabled?: boolean;
  controlled?: boolean;
  ariaLabel?: string;
  ariaLabelledBy?: string;
}

export class ToggleInput extends Component<ToggleInputProps> {
  render() {
    const {
      disabled = false,
      controlled,
      ariaLabel,
      ariaLabelledBy,
      checked,
      ...passProps
    } = this.props;
    return (
      <HtmlInput
        disabled={disabled}
        aria-label={ariaLabel}
        aria-labelledby={ariaLabelledBy}
        {...passProps}
        type="checkbox"
        checked={!!checked}
      />
    );
  }
}

export interface ToggleProps {
  /** Controlled toggle-state, use onClick to change  */
  checked?: boolean;
  /** Default status of toggle when not using controlled 'checked' state
   * @default false
   */
  defaultChecked?: boolean;
  /** Custom classname to extend or customize */
  className?: string;
  /** Disable Button usage */
  disabled?: boolean;
  /** Event handler to execute when clicked */
  onClick?: ({ toggleState }: { toggleState: boolean }) => void;
  /**
   * Label element content
   */
  children?: ReactNode;
  /** Pass custom props to Toggle's input component/element */
  toggleInputProps?: ToggleInputProps;
  /** Customized ToggleInput-component */
  toggleInputComponent?:
    | ComponentClass<ToggleInputProps>
    | FunctionComponent<ToggleInputProps>
    | ReactElement<ToggleInputProps>;
  /**
   * aria-label for the HTML input-element,
   * alternatively you can define aria-labelledby with label-element id
   */
  ariaLabel?: string;
  ariaLabelledBy?: string;
}

interface ToggleState {
  toggleState: boolean;
}

const componentOrElementWithProps = (
  component: ReactElement<any> | FunctionComponent<any> | ComponentClass<any>,
  props: object,
) => {
  if (!!component) {
    if (React.isValidElement(component)) {
      return React.cloneElement(component, props); // element
    }
    return React.createElement(
      component as FunctionComponent<any> | ComponentClass<any>,
      props,
    ); // component
  }
  return;
};

export class Toggle extends Component<ToggleProps> {
  state: ToggleState = {
    toggleState: !!this.props.checked || !!this.props.defaultChecked || false,
  };

  componentWillReceiveProps(nextProps: ToggleProps) {
    const { checked } = nextProps;
    if (!!checked) {
      this.setState({ toggleState: !!checked });
    }
  }

  handleClick = () => {
    const { checked, onClick } = this.props;
    const { toggleState } = this.state;
    if (checked === undefined) {
      this.setState({ toggleState: !toggleState });
    }
    if (!!onClick) {
      onClick({ toggleState: !toggleState });
    }
  };

  render() {
    const {
      disabled,
      className,
      children,
      toggleInputProps,
      toggleInputComponent,
      ariaLabel,
      ariaLabelledBy,
      checked: dissMissChecked,
      defaultChecked: dissMissDefaultChecked,
      onClick: dissMissOnClick,
      ...passProps
    } = this.props;
    const { toggleState } = this.state;
    const newToggleInputProps = {
      disabled,
      ariaLabel,
      ariaLabelledBy,
      checked: !!toggleState,
      className: toggleInputClassName,
      onChange: this.handleClick,
      ...toggleInputProps,
    };

    return (
      <HtmlSpan
        className={classnames(className, baseClassName, {
          [toggleDisabledClassName]: !!disabled,
        })}
      >
        {!!toggleInputComponent ? (
          componentOrElementWithProps(toggleInputComponent, newToggleInputProps) // element
        ) : (
          <ToggleInput {...newToggleInputProps} />
        )}
        <HtmlLabel
          {...passProps}
          className={toggleLabelClassName}
          onClick={this.handleClick}
          aria-labelledby="testaan"
        >
          {children}
        </HtmlLabel>
      </HtmlSpan>
    );
  }
}
