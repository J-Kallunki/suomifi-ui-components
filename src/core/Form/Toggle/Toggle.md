```js
import { Toggle } from 'suomifi-ui-components';

<Toggle onClick={({ toggleState }) => console.log(toggleState)}>
  Test
</Toggle>;
```
