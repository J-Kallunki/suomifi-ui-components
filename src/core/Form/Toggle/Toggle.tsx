import React, { Component } from 'react';
import styled from '@emotion/styled';
import classnames from 'classnames';
import { withDefaultTheme } from '../../theme/utils/defaultTheme';
import { ThemeComponent } from '../../theme';
import {
  baseStyles,
  inputBaseStyles,
  iconBaseStyles,
} from './Toggle.baseStyles';
import {
  Toggle as CompToggle,
  ToggleProps as CompToggleProps,
  ToggleInput,
  ToggleInputProps as CompToggleInputProps,
} from '../../../components/Form/Toggle';
import { Icon, IconProps } from '../../Icon/Icon';

export interface ToggleProps extends CompToggleProps, ThemeComponent {}
export interface ToggleInputProps
  extends CompToggleInputProps,
    ThemeComponent {}
interface ToggleIconProps extends IconProps, ThemeComponent {
  disabled: boolean;
  checked: boolean;
}

const baseClassName = 'fi-toggle';
const inputBaseClassName = `${baseClassName}-input`;
const iconBaseClassName = `${baseClassName}-icon`;

const StyledToggle = styled(
  ({ theme, className, ...passProps }: ToggleProps) => (
    <CompToggle
      {...passProps}
      className={classnames(className, {
        [`${baseClassName}--disabled`]: !!passProps.disabled,
      })}
    />
  ),
)`
  ${props => baseStyles(props)}
`;

const StyledInput = styled(({ theme, ...passProps }: ToggleInputProps) => (
  <ToggleInput {...passProps} className={inputBaseClassName} />
))`
  ${props => inputBaseStyles(props)}
`;

const StyledIcon = styled(
  ({ theme, className, disabled, checked, ...passProps }: ToggleIconProps) => (
    <Icon
      {...passProps}
      icon="toggle"
      className={classnames(className, {
        [`${iconBaseClassName}--disabled`]: !!disabled,
        [`${iconBaseClassName}--checked`]: !!checked,
      })}
    />
  ),
)`
  ${props => iconBaseStyles(props)}
`;

/**
 * Use for toggling form selection or application state
 */
export class Toggle extends Component<ToggleProps> {
  state = { toggleStatus: !!this.props.checked };

  handleToggle = () => {
    const { onClick } = this.props;
    const { toggleStatus } = this.state;
    this.setState({ toggleStatus: !toggleStatus });
    if (!!onClick) {
      onClick({ toggleState: !toggleStatus });
    }
  };

  render() {
    const {
      children,
      disabled = false,
      checked: dissMissChecked,
      onClick,
      ...passProps
    } = withDefaultTheme(this.props);
    const { toggleStatus } = this.state;
    const iconProps = {
      disabled,
      checked: toggleStatus,
      theme: passProps.theme,
    };
    return (
      <StyledToggle
        {...passProps}
        toggleInputComponent={<StyledInput theme={passProps.theme} />}
        onClick={this.handleToggle}
      >
        <StyledIcon {...iconProps} mousePointer={true} />
        {children}
      </StyledToggle>
    );
  }
}

export { ToggleInput };
