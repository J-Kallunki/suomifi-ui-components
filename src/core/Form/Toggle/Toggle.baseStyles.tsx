import { css } from '@emotion/core';
import { suomifiTheme } from '../../theme';
import { ToggleProps } from './Toggle';
import { element, font } from '../../theme/reset';

export const baseStyles = ({ theme = suomifiTheme }: ToggleProps) => css`
  ${element}
  ${font}
  background-color: ${theme.colors.white};
  .fi-toggle-label {
    cursor: pointer;
  }
`;

export const inputBaseStyles = ({ theme = suomifiTheme }: ToggleProps) => css`
  ${element}
  ${font}
  width: 0;
  height: 0;
  opacity: 0;
  z-index: -9999;
  background-color: ${theme.colors.white};
`;

export const iconBaseStyles = ({ theme = suomifiTheme }: ToggleProps) => css`
  width: 40px;
  height: 24px;
  margin-right: 8px;

  &.fi-toggle-icon--checked {
    .fi-toggle-icon-knob {
      transform: translateX(50%);
    }
    .fi-toggle-icon-slide {
      fill: ${theme.colors.toggleOnSlide};
    }
    .fi-toggle-icon-circle {
      fill: ${theme.colors.toggleOn};
    }
  }
`;
