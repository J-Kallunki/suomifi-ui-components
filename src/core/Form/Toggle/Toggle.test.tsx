import React from 'react';
import { render } from 'react-testing-library';
import { axeTest } from '../../../utils/test/axe';

import { Toggle } from './Toggle';

const doNothing = () => ({});

const TestToggleWithInput = (
  <Toggle
    onClick={doNothing}
    data-testid="toggle"
    id="test-toggle"
    withInput={true}
  >
    Test
  </Toggle>
);

test('calling render with the same component on the same container does not remount', () => {
  const buttonRendered = render(TestToggleWithInput);
  const { getByTestId, container, rerender } = buttonRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('toggle').textContent).toBe('Test');

  // re-render the same component with different props
  rerender(
    <Toggle onClick={doNothing} data-testid="elggot" withInput={true}>
      Test two
    </Toggle>,
  );
  expect(getByTestId('elggot').textContent).toBe('Test two');
});

test(
  'should not have basic accessibility issues',
  axeTest(TestToggleWithInput),
);
