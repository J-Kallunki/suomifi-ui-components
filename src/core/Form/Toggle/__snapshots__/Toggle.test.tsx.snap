// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-3 {
  margin: 0;
  padding: 0;
  border: 0;
  background: none;
  color: #282828;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  line-height: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  background-color: #FFFFFF;
}

.emotion-3 .fi-toggle-label {
  cursor: pointer;
}

.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  overflow: visible;
  box-sizing: border-box;
  line-height: 1;
  display: inline-block;
  max-width: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  background: none;
  color: #282828;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  line-height: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  width: 0;
  height: 0;
  opacity: 0;
  z-index: -9999;
  background-color: #FFFFFF;
}

.emotion-0::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.emotion-2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  box-sizing: border-box;
  line-height: 1;
  max-width: 100%;
}

.emotion-1 {
  max-width: 100%;
  display: inline-block;
  width: 24px;
  height: 24px;
  display: inline-block;
  vertical-align: text-bottom;
  width: 40px;
  height: 24px;
  margin-right: 8px;
}

.emotion-1 * {
  cursor: pointer;
}

.emotion-1.fi-toggle-icon--checked .fi-toggle-icon-knob {
  -webkit-transform: translateX(50%);
  -ms-transform: translateX(50%);
  transform: translateX(50%);
}

.emotion-1.fi-toggle-icon--checked .fi-toggle-icon-slide {
  fill: rgba(0,179,138,0.5);
}

.emotion-1.fi-toggle-icon--checked .fi-toggle-icon-circle {
  fill: #00B38A;
}

<span
  class="emotion-3 fi-toggle"
>
  <input
    class="fi-toggle-input emotion-0"
    id="testaan"
    type="checkbox"
  />
  <label
    aria-labelledby="testaan"
    class="fi-toggle-label emotion-2"
    data-testid="toggle"
  >
    <div
      aria-hidden="true"
      class="fi-icon emotion-1"
      focusable="false"
      role="img"
    >
      <div>
        <div
          class="fi-toggle-icon-svg"
          data-src="\\"toggle.svg\\""
          src=""
          style="fill:#003479;max-width:100%;max-height:100%;cursor:pointer"
        />
      </div>
    </div>
    Test
  </label>
</span>
`;
